amespahdbpythonsuite/transitions.py

```diff
@@ -543,7 +891,473 @@ def plot(self, **keywords) -> None:
+                * integrate.quad(Transitions.planck, 2.5e3, 1.1e5)[0]
+                / integrate.quad(Transitions.planck_number_of_photons, 2.5e3, 1.1e5)[0]
+            )
+
+        return me
+
+    @staticmethod
+    def mean_energy_squared(**keywords):
+        """
+        Callback function to calculate the mean energy in erg^2 for a given
+        blackbody temperatyre.
+
+        :Returns: float array
+
+        """
+        # ! Using simpson's integration, given there is no Python (scipy)
+        # ! implementation of the 5-point Newton-Cotes integration of
+        # ! IDL's INT_TABULATED.
+
+        if keywords.get('stellar_model'):
+            me = (
+                3.945966130997681e-32
+                * integrate.simpson(
+                    star_model['frequency'],
+                    star_model['frequency']
+                    * Transitions.absorption_cross_section(star_model['frequency'])
+                    * star_model['intensity'],
+                )
+                / integrate.simpson(
+                    star_model['frequency'],
+                    Transitions.absorption_cross_section(star_model['frequency'])
+                    * star_model['intensity']
+                    / star_model['frequency'],
+                )
+            )
+        elif keywords.get('isrf'):
+            me = (
+                3.945966130997681e-32
+                * integrate.quad(Transitions.isrf_squared, 2.5e3, 1.1e5)[0]
+                / integrate.quad(Transitions.isrf_number_of_photons, 2.5e3, 1.1e5)[0]
+            )
+        else:
+            me = (
+                3.945966130997681e-32
+                * integrate.quad(Transitions.planck_squared, 2.5e3, 1.1e5)[0]
+                / integrate.quad(Transitions.planck_number_of_photons, 2.5e3, 1.1e5)[0]
+            )
+
+        return me
+
+    @staticmethod
+    def attained_temperature(T: float) -> float:
+        """
+        Calculate a PAH's temperature after absorbing a given amount of energy.
+
+        :param T: Excitation temperature in Kelvin.
+        :type T: float
+
+        """
+        global energy
+
+        return integrate.quad(Transitions.heat_capacity, 2.73, T)[0] - energy
+
+    @staticmethod
+    def approximate_attained_temperature(T: float) -> float:
+        """
+        Calculate a PAH's temperature after absorbing a given amount of energy
+        using an approximation.
+
+        :param T: Excitation temperature in Kelvin.
+        :type T: float
+
+        """
+        global energy
+
+        return (
+            nc
+            * (
+                7.54267e-11 * special.erf(-4.989231 + 0.41778 * np.log(T))
+                + 7.542670e-11
+            )
+            - energy
+        )
+
+    @staticmethod
+    def heat_capacity(T: float) -> float:
+        """
+        Calculate heat capacity.
+
+        :param T: Excitation temperature in Kelvin.
+        :type T: float
+
+        """
+
+        global frequencies
```

Comment:
Will need to find a mechanism to not use globals; either via a static class variable or using partials.